src/components/key-capturer/key-capturer.tsx

```diff
@@ -135,54 +135,71 @@ type KeyCapturerProps = EventHandlerProps &
-    const composingRef = React.useRef(false)
-    const composingEventHandlers = props.onEnter
-        ? {
-              onCompositionStart: () => {
-                  composingRef.current = true
-              },
-              onCompositionEnd: () => {
-                  composingRef.current = false
-              },
-          }
-        : undefined
-
-    function handleKeyEvent(event: React.KeyboardEvent<HTMLInputElement>) {
-        // https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode
-        const key =
-            event.key !== undefined
-                ? KeyCapturerResolver.resolveByKey(event.key)
-                : KeyCapturerResolver.resolveByKeyCode(event.keyCode)
-
-        if (!key) return
-        const propagateEvent = props[keyPropagatePropMapping[key]] || false
-        const eventHandler = props[keyEventHandlerMapping[key]]
-
-        if (key === 'Enter' && eventHandler) {
-            if (
-                composingRef.current ||
-                // Safari fires the onCompositionEnd event before the keydown event, so we
-                // have to rely on the 229 keycode, which is Enter when fired from an IME
-                // https://www.w3.org/TR/uievents/#determine-keydown-keyup-keyCode
-                (event.keyCode || event.which) === 229
-            ) {
-                return
-            }
-        }
-
-        if (eventHandler) {
-            eventHandler(event)
-            if (!propagateEvent) {
-                event.preventDefault()
-                event.stopPropagation()
-            }
-        }
-    }
-
-    return React.cloneElement(children, {
-        [eventName]: handleKeyEvent,
-        ...composingEventHandlers,
-    })
+    const [isComposing, setIsComposing] = React.useState(false)
+
+    const onCompositionStart = React.useMemo(
+        () =>
+            props.onEnter
+                ? () => {
+                      setIsComposing(true)
+                  }
+                : undefined,
+        [props.onEnter],
+    )
+    const onCompositionEnd = React.useMemo(
+        () =>
+            props.onEnter
+                ? () => {
+                      setIsComposing(false)
+                  }
+                : undefined,
+        [props.onEnter],
+    )
+
+    const handleKeyEvent = React.useMemo(
+        () =>
+            function handleKeyEvent(event: React.KeyboardEvent<HTMLInputElement>) {
+                // https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode
+                const key =
+                    event.key !== undefined
+                        ? KeyCapturerResolver.resolveByKey(event.key)
+                        : KeyCapturerResolver.resolveByKeyCode(event.keyCode)
+
+                if (!key) return
+                const propagateEvent = props[keyPropagatePropMapping[key]] || false
+                const eventHandler = props[keyEventHandlerMapping[key]]
+
+                if (key === 'Enter' && eventHandler) {
```

Comment:
Given we're checking again for `eventHandler` on line 184, we could probably invert this check and check for `eventHandler` first, then whether `key` is enter:



```
if(eventHandler) {

	if(key === 'Enter' & ( isComposing ||
	  // Safari fires the onCompositionEnd event before the keydown event, so we
	  // have to rely on the 229 keycode, which is Enter when fired from an IME
	  // https://www.w3.org/TR/uievents/#determine-keydown-keyup-keyCode
	  (event.keyCode || event.which) === 229)
	) {
	
		return 
	}


	eventHandler(event)
     if (!propagateEvent) {
          event.preventDefault()
          event.stopPropagation()
      }
}

```